src/storage/sql/mod.rs

```diff
@@ -1 +1,319 @@
+mod models;
 mod schema;
+
+use std::convert::From;
+
+use diesel::pg::upsert;
+use diesel::prelude::*;
+use diesel::r2d2::{ConnectionManager, Pool, PoolError};
+use diesel::result::{DatabaseErrorKind, Error::DatabaseError};
+
+use super::{errors::StorageError, Snippet, Storage};
+use schema::{changesets, snippets, tags};
+
+/// A Storage implementation which persists snippets' data in a SQL database.
+pub struct SqlStorage {
+    pool: Pool<ConnectionManager<PgConnection>>,
+}
+
+impl SqlStorage {
+    pub fn new(database_url: &str) -> Result<SqlStorage, StorageError> {
+        let manager = ConnectionManager::new(database_url);
+        let pool = Pool::builder().build(manager)?;
+
+        Ok(Self { pool })
+    }
+
+    fn upsert_snippet(&self, conn: &PgConnection, snippet: &Snippet) -> Result<i32, StorageError> {
+        let result = diesel::insert_into(snippets::table)
+            .values((
+                snippets::slug.eq(&snippet.id),
+                snippets::title.eq(&snippet.title),
+                snippets::syntax.eq(&snippet.syntax),
+                snippets::created_at.eq(diesel::dsl::now),
+            ))
+            .on_conflict(snippets::slug)
+            .do_update()
+            .set((
+                snippets::title.eq(&snippet.title),
+                snippets::syntax.eq(&snippet.syntax),
+                snippets::updated_at.eq(diesel::dsl::now),
+            ))
+            .returning(snippets::id)
+            .get_result::<i32>(conn);
+        let snippet_id = match result {
+            Ok(snippet_id) => snippet_id,
+            Err(DatabaseError(DatabaseErrorKind::UniqueViolation, _)) => {
+                return Err(StorageError::Duplicate {
+                    id: snippet.id.to_owned(),
+                })
+            }
+            Err(e) => return Err(StorageError::from(e)),
+        };
```

Comment:
If it's `upsert`, in which case it may return a duplicate error?

Reply:
Sorry, this I forgot to push the updated commit :(